deployer/src/builder/oci/image/mod.rs

```diff
@@ -0,0 +1,268 @@
+///
+/// ```text
+/// ghcr.io/termoshtt/ocipkg/testing:latest
+/// ^^^^^^^---------------------------------- hostname
+///         ^^^^^^^^^^^^^^^^^^^^^^^^--------- name
+///                                  ^^^^^^-- reference
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ImageName {
+    pub hostname: String,
+    pub port: Option<u16>,
+    pub name: Name,
+    pub reference: Reference,
+}
+
+impl fmt::Display for ImageName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(port) = self.port {
+            write!(
+                f,
+                "{}:{}/{}:{}",
+                self.hostname, port, self.name, self.reference
+            )
+        } else {
+            write!(f, "{}/{}:{}", self.hostname, self.name, self.reference)
+        }
+    }
+}
+
+impl ImageName {
+    pub fn parse(name: &str) -> Result<Self> {
+        let (hostname, name) = name
+            .split_once('/')
+            .ok_or(super::error::Error::InvalidName(
+                "Couldn't get the hostname of the image".to_string(),
+            ))?;
+        let (hostname, port) = if let Some((hostname, port)) = hostname.split_once(':') {
+            (hostname, Some(str::parse(port)?))
+        } else {
+            (hostname, None)
+        };
+        let (name, reference) = name.split_once(':').unwrap_or((name, "latest"));
+        Ok(ImageName {
+            hostname: hostname.to_string(),
+            port,
+            name: Name::new(name)?,
+            reference: Reference::new(reference)?,
+        })
+    }
+
+    /// URL for OCI distribution API endpoint
+    pub fn registry_url(&self) -> Result<Url> {
+        let hostname = if let Some(port) = self.port {
+            format!("{}:{}", self.hostname, port)
+        } else {
+            self.hostname.clone()
+        };
+        let url = if self.hostname.starts_with("localhost") {
+            format!("http://{}", hostname)
+        } else {
+            format!("https://{}", hostname)
+        };
+        Ok(Url::parse(&url)?)
+    }
+}
+
+/// Handler for oci-archive format
+///
+/// oci-archive consists of several manifests i.e. several container.
+pub struct Archive<'a> {
+    inner: &'a [u8],
+    archive: Option<tokio_tar::Archive<&'a [u8]>>,
+}
+
+impl<'a> Archive<'a> {
+    pub fn new(buf: &'a [u8]) -> Self {
+        Self {
+            inner: buf,
+            archive: None,
+        }
+    }
+
+    pub fn entries(&mut self) -> Result<tokio_tar::Entries<&'a [u8]>> {
+        // Calling entries must be preceded by a reseted archive, which is denoted
+        // by a non empty Option. Consumers of the archive, meaning scopes where
+        // `.entries()` is called will take out the value from the Option leaving
+        // it empty for the next `.entries()` calls.
+        if self.archive.is_some() {
+            let mut inner = self.archive.take().expect("to get the inner archive");
```

Comment:
Should we have proper error handling here too?

Reply:
Updated how this gets done. Our calls to `.entries()` must be preceded by the `self.archive` reinitialization, so I simply moved the reinit to the `.entries()` body, and now the `.expect(...)` call is justified. I don't think we'll need to handle an error anymore given how we construct this.